opentelemetry/api/trace/__init__.py

```diff
@@ -0,0 +1,190 @@
+API package alone without a supporting implementation.
+
+The tracer supports implicit and explicit context propagation. By default spans
+are created as children of the currently active span, and the newly-created
+span becomes the new active span::
+
+    from opentelemetry.sdk.trace import tracer
+
+    # Create a new root span, set it as the current span in context
+    with tracer.span("parent"):
+        # Attach a new child and update the current span
+        with tracer.span("child"):
+            do_work():
+        # Close child span, set parent as current
+    # Close parent span, set default span as current
+
+Under explicit context propagation there is no concept of an active span, and
+the caller is responsible for managing the span's lifetime::
+
+    from opentelemetry.sdk.trace import tracer
+    from your.integration import deserialize_span
+
+    parent = deserialize_span(serialized_span)
+    # Explicit parent span assignment
+    with tracer.span("child", parent=parent) as child:
+        do_work(span=child)
+
+Applications should generally use a single global tracer, and use either
+implicit or explicit context propagation consistently throughout.
+
+.. versionadded:: 0.1.0
+"""
+
+from __future__ import annotations
+
+from typing import Iterator
+from contextlib import contextmanager
+
+
+class Tracer(object):
+    """Handles span creation and in-process context propagation.
+
+    This class provides methods for manipulating the context, creating spans,
+    and controlling spans' lifecycles.
+    """
+
+    def get_current_span(self) -> Span:
+        """Get the currently active span from the context.
+
+        Returns:
+            The currently active :class:`.Span`.
+        """
+        raise NotImplementedError
+
+    @contextmanager
+    def span(self, name: str, parent: Span=None) -> Iterator[None]:
+        """Context manager for span creation.
+
+        Create a new child of the current span, or create a root span if no
+        current span exists. Start the span and set it as the current span in
+        this tracer's context.
+
+        On exiting the context manager stop the span and set its parent as the
+        current span.
+
+        Example::
+
+            with tracer.span("one") as parent:
+                parent.add_event("parent's event")
+                with tracer.span("two") as child:
+                    child.add_event("child's event")
+                    tracer.get_current_span()  # returns child
+                tracer.get_current_span()      # returns parent
+            tracer.get_current_span()          # returns the previously active span
+
+        Args:
+            name: The name of the span to be created.
+            parent: This span's parent.
+
+        Raises:
+            ValueError: if ``name`` is null.
+        """
+        raise NotImplementedError
+
+    def create_span(self, name: str, parent: Span=None) -> Span:
```

Comment:
How to create a root span which has no parent?

Reply:
True. In fact, I think its strange that `create_span(..., parent=None)` should have a non-None parent. One suggestion: `create_span` should just never affect or use (as parent) the active span. But instead, enhance the context manager returned by `span` to return the created span at `__enter__`.

Reply:
I can think of a few options here: use a sentinel default value instead of null and take `parent=None` to mean that we should create a root span, add a separate method `create_root_span`, or change the way we create spans and match java's builder pattern more closely. 

> `create_span` should just never affect or use (as parent) the active span

That's pretty surprising, I'd expect creating a child of the current active span to be the most common use case.

> But instead, enhance the context manager returned by `span` to return the created span at `__enter__`

That's what I'd expect the context manager to do now.